### PR TITLE
fix(electrostatics): handle q(R) charge gradients

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Fix electrostatics autograd routing for geometry-dependent charges, where `charges = q(positions)` can otherwise combine connected position gradients and explicit charge gradients from the same recompute path. The updated Ewald and PME Torch wrappers route q(R) workloads through charge-gradient injection or fallback recomputes that separate fixed-charge geometry partials from `dE/dq`, letting PyTorch apply the `dE/dq * dq/dR` chain term once.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes https://github.com/NVIDIA/nvalchemi-toolkit-ops/issues/115

## Changes Made

- Add a shared q(R) detection helper and use it in cached Ewald/PME Torch autograd paths.
- Recompute safe fixed-charge geometry partials for non-uniform cotangents and higher-order fallback paths while returning charge gradients separately.
- Route direct-output q(R) Ewald/PME paths through charge-gradient injection where positions and cells are detached.
- Add q(R) manual-chain, weighted-loss, HVP, hybrid-force, and batched double-backward regression coverage for Ewald and PME.
- Rename legacy direct-output wording to deprecated direct-output wording in touched electrostatics code and tests.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

Not run while preparing PR materials.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

None.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/torch/interactions/electrostatics/_util.py [hunk 1]
```diff
@@ -46,6 +47,37 @@ def _sum_charge_gradients(
...
+    Issue #115 is prevented by routing these workloads through
+    :class:`_InjectChargeGrad`, eager electrostatics chains, or safe cached
+    fallback paths that recompute independent partial derivatives before PyTorch
+    applies the dE/dq * dq/dR chain term exactly once.
+    """
+    return bool(
+        positions.requires_grad
+        and charges.requires_grad
+        and charges.grad_fn is not None
+    )
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/_util.py:78]: Known limitation.

File: nvalchemiops/torch/interactions/electrostatics/ewald.py [hunk 2]
```diff
@@ -806,24 +821,33 @@ def _fallback(p, q, c):
...
-                batch_idx,
-                _fallback,
-            )
+            if charge_graph_depends_on_positions:
+                energies = _InjectChargeGrad.apply(
+                    energies,
+                    charges,
+                    charge_grads.detach(),
+                    batch_idx,
+                )
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/ewald.py]: Fixed in 19e9267eb60e1480cd582bbe9698699368676323.

File: examples/electrostatics/02_ewald_summation_example.py [hunk 3]
```diff
@@ -494,6 +496,7 @@ def create_nacl_system(n_cells: int = 2, lattice_constant: float = 5.64):
 cells_batch = torch.cat(all_cells, dim=0)
 pbc_batch = torch.cat(all_pbc, dim=0)
 batch_idx = torch.tensor(batch_idx_list, dtype=torch.int32, device=device)
+max_atoms_per_system = max(atom_counts)
```
Comment (dallasfoster) [inline examples/electrostatics/02_ewald_summation_example.py:500]: Is this not a property of Batch? Maybe it should be?

File: nvalchemiops/torch/interactions/electrostatics/ewald.py [hunk 4]
```diff
@@ -243,6 +244,26 @@ def _atom_ranges(
...
+    value = int(max_atoms_per_system)
+    if positions.shape[0] == 0:
+        if value < 0:
+            raise ValueError("max_atoms_per_system must be non-negative")
+        return value
+    if value <= 0:
+        raise ValueError("max_atoms_per_system must be positive for non-empty batches")
+    if value > positions.shape[0]:
+        raise ValueError("max_atoms_per_system cannot exceed the total number of atoms")
+    return value
```
Comment (dallasfoster) [inline nvalchemiops/torch/interactions/electrostatics/ewald.py]: Why do we need this function and the similar function in _ewald_recip_chain.py?

File: nvalchemiops/torch/interactions/electrostatics/ewald.py [hunk 5]
```diff
@@ -603,6 +625,9 @@ def _reciprocal_space_energy(
         )
         volume = torch.abs(torch.linalg.det(cell)).to(torch.float64)
         atom_start, atom_end = _atom_ranges(batch_idx, num_systems)
+        max_atoms_hint = _normalize_max_atoms_per_system(
```
Comment (dallasfoster) [inline nvalchemiops/torch/interactions/electrostatics/ewald.py]: I don't really like the "_hint" designation here because it's not actually a hint, right?

File: nvalchemiops/torch/interactions/electrostatics/_util.py [hunk 6]
```diff
@@ -46,6 +47,37 @@ def _sum_charge_gradients(
     return real_space_charge_grads + reciprocal_charge_grads
 
 
+def _has_potentially_geometry_dependent_charges(
+    positions: torch.Tensor,
+    charges: torch.Tensor,
+) -> bool:
```
Comment (dallasfoster) [inline nvalchemiops/torch/interactions/electrostatics/_util.py:53]: I don't really understand why this has to be it's own function and not just an inline piece of code.

Comment (dallasfoster) [reply nvalchemiops/torch/interactions/electrostatics/_util.py:53]: I guess I could go either way on this one

**Review by dallasfoster — CHANGES_REQUESTED** [review]
Just a couple small comments

File: nvalchemiops/torch/interactions/electrostatics/_util.py (hunk 6 above)
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/_util.py:53]: Shared by real-space Ewald, cached fallback, and PME reciprocal backward paths.

File: nvalchemiops/torch/interactions/electrostatics/ewald.py (hunk 5 above)
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/ewald.py]: Addressed in d51e46f0c0835b38db64df1c477600fef70af3cc.

File: nvalchemiops/torch/interactions/electrostatics/ewald.py (hunk 4 above)
Comment (zubatyuk) [reply nvalchemiops/torch/interactions/electrostatics/ewald.py]: Addressed in d51e46f0c0835b38db64df1c477600fef70af3cc.
